simulator/src/scrypto_bindgen/translation.rs

```diff
@@ -0,0 +1,310 @@
+        TypeKind::U16 => type_ident.unwrap_or("u16".to_owned()),
+        TypeKind::U32 => type_ident.unwrap_or("u32".to_owned()),
+        TypeKind::U64 => type_ident.unwrap_or("u64".to_owned()),
+        TypeKind::U128 => type_ident.unwrap_or("u128".to_owned()),
+        TypeKind::String => type_ident.unwrap_or("String".to_owned()),
+        TypeKind::Array { element_type } => type_ident.unwrap_or(format!(
+            "Vec<{}>",
+            type_name(
+                &ScopedTypeId(type_identifier.0, element_type),
+                blueprint_name,
+                schema_resolver
+            )?
+        )),
+        TypeKind::Tuple { field_types } => type_ident.unwrap_or(if field_types.is_empty() {
+            "()".to_owned()
+        } else {
+            format!(
+                "({},)",
+                field_types
+                    .iter()
+                    .map(|local_type_index| type_name(
+                        &ScopedTypeId(type_identifier.0, *local_type_index),
+                        blueprint_name,
+                        schema_resolver
+                    ))
+                    .collect::<Result<Vec<String>, _>>()?
+                    .join(", ")
+            )
+        }),
+        TypeKind::Enum { variants } => {
+            // There is currently no way to know if this type has generics or not. Thus, we need to
+            // deal with generic enums from the standard library in a special way. We determine if
+            // the type at hand is `Option` if: the name in the metadata is "Option" and the local
+            // type index is of a well known type.
+            match (
+                type_metadata.get_name(),
+                variants.len(),
+                variants.get(&0).as_ref().map(|vec| vec.as_slice()),
+                variants.get(&1).as_ref().map(|vec| vec.as_slice()),
+            ) {
+                (Some("Option"), 2usize, Some([]), Some([some_type_index])) => Ok(format!(
+                    "Option<{}>",
+                    type_name(
+                        &ScopedTypeId(type_identifier.0, *some_type_index),
+                        blueprint_name,
+                        schema_resolver
+                    )?
+                )),
+                (Some("Result"), 2usize, Some([ok_type_index]), Some([err_type_index])) => {
+                    Ok(format!(
+                        "Result<{}, {}>",
+                        type_name(
+                            &ScopedTypeId(type_identifier.0, *ok_type_index),
+                            blueprint_name,
+                            schema_resolver
+                        )?,
+                        type_name(
+                            &ScopedTypeId(type_identifier.0, *err_type_index),
+                            blueprint_name,
+                            schema_resolver
+                        )?
+                    ))
+                }
+                (Some(name), ..) => Ok(name.to_owned()),
```

Comment:
Question: This is the reason why the schema type name must match the Rust type name?

Reply:
Yup, exactly! It's this bit and also bits like:

```rust
type_ident.unwrap_or(if field_types.is_empty() {
    "()".to_owned()
} else {
    format!(
        "({},)",
        field_types
            .iter()
            .map(|local_type_index| type_name(
                &ScopedTypeId(type_identifier.0, *local_type_index),
                schema_resolver
            ))
            .collect::<Result<Vec<String>, _>>()?
            .join(", ")
    )
})
```

where we prefer returning the `type_ident` that we got from the schema metadata over the types that we construct. So, there is a name in the metadata schema which is not to a rust type we end up returning that and getting a compile error in the final stubs since that type doesn't exist.